File: AI_LAB3/EX10/Nhom2_28_B2113306_KimKhanhDang_8queens.py
# In ra mot trong cac trang thai kha thi

def is_safe(board, row, col):
    for i in range(row):
        if board[i][col] == 1:
            return False

    for i, j in zip(range(row, -1, -1), range(col, -1, -1)):
        if board[i][j] == 1:
            return False

    for i, j in zip(range(row, -1, -1), range(col, len(board))):
        if board[i][j] == 1:
            return False

    return True

def solve_n_queens(board, row):
    if row == len(board):
        print_solution(board)
        return True

    for col in range(len(board)):
        if is_safe(board, row, col):
            board[row][col] = 1
            if solve_n_queens(board, row + 1):
                return True
            board[row][col] = 0

def print_solution(board):
    for row in board:
        print(" ".join(["Q" if cell == 1 else "." for cell in row]))
    print("\n")
n = 8
chessboard = [[0 for _ in range(n)] for _ in range(n)]

if __name__ == '__main__':
    if not solve_n_queens(chessboard, 0):
        print('NONE SOLUTION')


# # In ra tat ca cac trang thai kha thi
# def is_safe(board, row, col):
#     for i in range(row):
#         if board[i][col] == 1:
#             return False
#
#     for i, j in zip(range(row, -1, -1), range(col, -1, -1)):
#         if board[i][j] == 1:
#             return False
#
#     for i, j in zip(range(row, -1, -1), range(col, len(board))):
#         if board[i][j] == 1:
#             return False
#
#     return True
#
# def solve_n_queens(board, row):
#     if row == len(board):
#         print_solution(board)
#         return True
#
#     for col in range(len(board)):
#         if is_safe(board, row, col):
#             board[row][col] = 1
#             solve_n_queens(board, row + 1)
#             board[row][col] = 0
#
# def print_solution(board):
#     for row in board:
#         print(" ".join(["Q" if cell == 1 else "." for cell in row]))
#     print("\n")
# n = 8
# chessboard = [[0 for _ in range(n)] for _ in range(n)]
# solve_n_queens(chessboard, 0)

File: AI_LAB3/EX10/Nhom2_28_B2113306_KimKhanhDang_8queens.c
#include <stdio.h>
#include <stdbool.h>

#define N 8

void printBoard(int board[N][N]) {
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            if (board[i][j] == 1) {
                printf("Q ");
            } else {
                printf(". ");
            }
        }
        printf("\n");
    }
}

bool isSafe(int board[N][N], int row, int col) {
    for (int i = 0; i < row; i++) {
        if (board[i][col])
            return false;
    }

    for (int i = row, j = col; i >= 0 && j >= 0; i--, j--) {
        if (board[i][j])
            return false;
    }

    for (int i = row, j = col; i >= 0 && j < N; i--, j++) {
        if (board[i][j])
            return false;
    }

    return true;
}

bool solveNQueens(int board[N][N], int row) {
    if (row == N) {
        return true;
    }

    for (int col = 0; col < N; col++) {
        if (isSafe(board, row, col)) {
            board[row][col] = 1;
            if (solveNQueens(board, row + 1))
                return true;
            board[row][col] = 0;
        }
    }
    return false;
}

int main() {
    int board[N][N] = {0};

    if (solveNQueens(board, 0)) {
        printf("SOLUTION:\n");
        printBoard(board);
    } else {
        printf("NONE SOLUTION.\n");
    }

    return 0;
}

